xarray_beam/_src/pangeo.py

```diff
@@ -0,0 +1,150 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""IO with Pangeo-Forge."""
+from typing import (
+  Dict,
+  Iterator,
+  Optional,
+  Mapping,
+  Tuple,
+  cast,
+)
+
+import apache_beam as beam
+import xarray
+from apache_beam.io.filesystems import FileSystems
+
+from xarray_beam._src import core
+
+
+def _zero_dimensions(dataset: xarray.Dataset) -> Mapping[str, int]:
+  return {dim: 0 for dim in dataset.dims.keys()}
+
+
+def _expand_dimensions_by_key(
+    dataset: xarray.Dataset,
+    key: core.Key,
+    index: Tuple[int, ...],
+    pattern: 'FilePattern'
+) -> xarray.Dataset:
+  """Expand the dimensions of the `Dataset` by offsets found in the `Key`."""
+  combine_dims_by_name = {
+    combine_dim.name: (i, combine_dim)
+    for i, combine_dim in enumerate(pattern.combine_dims)
+  }
+
+  if not combine_dims_by_name:
+    return dataset
+
+  for dim_key in key.offsets.keys():
+    # skip expanding dimensions if they already exist
+    if dim_key in dataset.dims:
+      continue
+
+    dim_idx, combine_dim = combine_dims_by_name.get(dim_key, (-1, None))
+    if dim_idx == -1:
+      raise ValueError(
+        f"could not find CombineDim named {dim_key!r} in pattern {pattern!r}."
+      )
+
+    dim_val = combine_dim.keys[index[dim_idx]]
+    dataset = dataset.expand_dims(**{dim_key: [dim_val]})
+
+  return dataset
+
+
+class FilePatternToChunks(beam.PTransform):
+  """Open data described by a Pangeo-Forge `FilePattern` into keyed chunks."""
+
+  from pangeo_forge_recipes.patterns import FilePattern
+
+  def __init__(
+      self,
+      pattern: 'FilePattern',
+      sub_chunks: Optional[Mapping[str, int]] = None,
+      xarray_open_kwargs: Optional[Dict] = None
+  ):
+    """Initialize FilePatternToChunks.
+
+    TODO(#29): Currently, `MergeDim`s are not supported.
+
+    Args:
+      pattern: a `FilePattern` describing a dataset.
+      sub_chunks: split each open dataset into smaller chunks. If not set, each
+        chunk will open the full dataset.
+      xarray_open_kwargs: keyword arguments to pass to `xarray.open_dataset()`.
+    """
+    self.pattern = pattern
+    self.sub_chunks = sub_chunks or -1
+    self.xarray_open_kwargs = xarray_open_kwargs or {}
+
+  def _prechunk(self) -> Iterator[Tuple[core.Key, Tuple[int, ...]]]:
+    """Converts `FilePattern` items into keyed indexes."""
+    input_chunks = {k: v or 1 for k, v in self.pattern.nitems_per_input.items()}
```

Comment:
Are you sure `v or 1` is correct? it seems like `v=None` would correspond to "unknown" size, which would need to be inferred (by looking at the dataset) in order to compute offsets.

Reply:
I will give more thought to this, especially given #32 -- since I wrote this code, it seems like Pangeo-Forge indexes have changed.